src/utils/scopes.js

```diff
@@ -113,6 +122,233 @@ function getThisVariable(frame: any, path: string) {
+        : L10N.getStr("anonymous");
+  } else {
+    title = L10N.getStr("scopes.block");
+  }
+
+  let vars = [];
+  bindingsNames.forEach(name => {
+    // Find binding name in the original source bindings
+    const generatedScope = generatedScopes.find(
+      gs => gs.sourceBindings && name in gs.sourceBindings
+    );
+    if (!generatedScope || !generatedScope.sourceBindings) {
+      return;
+    }
+    // .. and map it to the generated name
+    const generatedName = generatedScope.sourceBindings[name];
+    // Skip if we already use the generated name
+    if (generatedName && !foundGeneratedNames[generatedName]) {
+      if (generatedScope.bindings.variables[generatedName]) {
+        vars.push({
+          name,
+          generatedName,
+          path: `${key}/${generatedName}`,
+          contents: generatedScope.bindings.variables[generatedName]
+        });
+        foundGeneratedNames[generatedName] = true;
+        return;
+      }
+
+      const arg = generatedScope.bindings.arguments.find(
+        arg_ => arg_[generatedName]
+      );
+      if (arg) {
+        vars.push({
+          name,
+          generatedName,
+          path: `${key}/${generatedName}`,
+          contents: arg[generatedName]
+        });
+        foundGeneratedNames[generatedName] = true;
+        return;
+      }
+    }
+
+    vars.push({
+      name,
+      generatedName,
+      path: `${key}/${generatedName}`,
+      contents: { value: { type: "undefined" } }
+    });
+  });
+
+  if (isLast) {
+    // For the last synthesized scope, apply all generated names we did not use
+    const allGeneratedVars = generatedScopes.reduce((acc_, { bindings }) => {
+      return acc_.concat(getBindingVariables(bindings, key));
+    }, []);
+    vars = vars.concat(
+      allGeneratedVars.filter(v => !foundGeneratedNames[v.name])
+    );
+  }
+
+  if (index === 0) {
+    // For the first synthesized scope, add this and other vars.
+    if (pauseInfo) {
+      vars = vars.concat(getSpecialVariables(pauseInfo, key));
+    }
+
+    if (selectedFrame) {
+      const this_ = getThisVariable(selectedFrame, key);
+
+      if (this_) {
+        vars.push(this_);
+      }
+    }
+  }
+
+  if (vars && vars.length) {
+    vars.sort((a, b) => a.name.localeCompare(b.name));
+    result.push({
+      name: title,
+      path: key,
+      contents: vars
+    });
+  }
+  return acc;
+}
+
+function synthesizeScopes(
+  scope: Scope,
+  selectedFrame: ?Frame,
+  pauseInfo: ?Pause,
+  scopeIndex: number,
+  scopes
+): NamedValue[] {
+  const { actor, syntheticScopes } = scope;
+  if (!syntheticScopes) {
+    return [];
+  }
+
+  // Collect all connected generated scopes.
+  const generatedScopes = [];
+  for (
+    let count = syntheticScopes.groupLength, s = scope;
+    count > 0 && s;
+    count--
+  ) {
+    generatedScopes.push(s);
+    s = s.parent;
+  }
+
+  const { result } = syntheticScopes.scopes.reduce(synthesizeScope, {
```

Comment:
I think the fact that synthesizeScope is mutating makes synthesize a bit more complicated... perhaps, we could have it return the bindings...

```js
  const { result } = syntheticScopes.scopes.reduce((bloby, syntheticScope, index) => {
    const bindings = getBindingsFromScope(bloby,..) 
    // it's tempting to update foundGeneratedNames here too so that `getBindingsFromScope` is pure, 
    // but not sure if that is needed... 
    if (bindings.length) {
      const title = getScopeTitle(...);
      blob.result.push({
        name: title,
        path: key,
        contents: vars
      });
    }
  }, blobOStuff)

```